### PR TITLE
Experimental support of jemalloc on non-linux

### PR DESCRIPTION
**Pull request by azat**
Supports:
- [x] freebsd (works)
- [x] ~darwin~ (disabled due to some issues)

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Cc: @alexey-milovidov 
Cc: @akuzm 

<details>

HEAD:
- f5d0dbde7182dea725c32508be9c3c1405ce5b49
- f06af22c7bd04e1fbcf6e05f1fbbc9d65a395f8c
- 3475baaaec
- 51d2f436b658fd1239d3acb9ea601b63c80e154c
- 8b6e2e6c67fab32cf599a0f2f8bcddc9326d293e

</details>

### CODE REVIEW

**Comment by azat** [comment]
@alexey-milovidov any interest in this?

I'm expecting that after the last patch the build will succeed, I can rebase (+ maybe add more comments)

**Comment by alexey-milovidov** [comment]
I'm interested in this pull request. It is better to use the same allocator on all systems.

BTW, do you have setups to test it? I'm asking because we don't run tests on non-Linux.

ClickHouse is used on Mac for two use cases:
- for clickhouse-client;
- for development.

ClickHouse is also used on FreeBSD servers but until recently people use some old version from "ports": https://www.freshports.org/databases/clickhouse/

**Comment by azat** [comment]
>BTW, do you have setups to test it?

No, I started looking into this just because noticed some issues in generated headers for jemalloc

**Comment by azat** [comment]
>ClickHouse build check — 10/18 builds are OK

Some CI internal issues, build on freebsd passed, so everything should work, but requires last rebase

**Comment by azat** [comment]
@alexey-milovidov BTW it is disabled by default for now, but can be manually enabled (while before you cannot even enable it), since I did not checked the result

Once someone will report that it works, it can be enabled by default

**Comment by alexey-milovidov** [comment]
You can run built binary in Vagrant/Virtualbox/Qemu and test that it does not crash on trivial queries.

**Comment by azat** [comment]
>You can run built binary in Vagrant/Virtualbox/Qemu and test that it does not crash on trivial queries.

I'm not the fan of freebsd, but I completely forgot that I can just download the binary from CI, I guess this I can do

**Comment by azat** [comment]
>You can run built binary in Vagrant/Virtualbox/Qemu and test that it does not crash on trivial queries.

Yep, that was a good idea, right now - it does not works (at least for freebsd)

**Comment by azat** [comment]
@alexey-milovidov what do you think, should it be enabled by default for freebsd (once it will work) ? Or the behavior that the PR has right now is more suitable (disable by default, but can be enabled with `-DENABLE_JEMALLOC=1`) until more testing will done ?

**Comment by alexey-milovidov** [comment]
@azat It's better to make it by default, because ClickHouse will work bettter. If I remember correctly, FreeBSD is also using jemalloc by default but the version will be older and without patches for performance.

**Comment by azat** [comment]
>@azat It's better to make it by default, because ClickHouse will work bettter.

Ok, actually I already did this (with experimental message just in case) in the latest version (replace ifdef in headers with separate header for each os/arch)

**Comment by azat** [comment]
>ClickHouse build check — Some exception occurred :(

Will it be restarted? (I saw the same issue in another PR)

**Comment by alexey-milovidov** [comment]
> Will it be restarted?

No. You can make another commit to restart CI checks.

**Comment by azat** [comment]
Now for non-linux setup it has a warning, but it uses number of available CPUs correctly on darwin/freebsd:

> <jemalloc>: No getcpu support: percpu_arena:percpu
> <jemalloc>: perCPU arena getcpu() not available. Setting narenas to 1.

**Comment by azat** [comment]
>I'm interested in this pull request. It is better to use the same allocator on all systems.

@alexey-milovidov freebsd is supported now (simple check passed), but I disabled jemalloc for darwin for now (since there are some issues - f76511668f3bda374f66d1176cc9d85d3259e117).

BTW any thoughts on adding binaries with debug symbols for freebsd/darwin?

**Comment by alexey-milovidov** [comment]
@azat I don't know why they are built without debug information.
We can enable it any time.

File: contrib/jemalloc-cmake/include_linux_aarch64/jemalloc/internal/jemalloc_internal_defs.h.in [hunk 1]
```diff
@@ -35,7 +35,7 @@
  */
 #define CPU_SPINWAIT
 /* 1 if CPU_SPINWAIT is defined, 0 otherwise. */
-#define HAVE_CPU_SPINWAIT 9
+#define HAVE_CPU_SPINWAIT 0
```
Comment (alexey-milovidov) [inline contrib/jemalloc-cmake/include_linux_aarch64/jemalloc/internal/jemalloc_internal_defs.h.in:38]: I wonder what is the reason for this change...

Comment (azat) [reply contrib/jemalloc-cmake/include_linux_aarch64/jemalloc/internal/jemalloc_internal_defs.h.in:38]: You should wonder before, in https://github.com/ClickHouse/ClickHouse/commit/82f2b1a986965fc850188460f4873566101692a0#diff-ee00bf922b046bbd74fef9b9dd42e366L44-R38 :)
This is a fix for a typo that I made in #11163, this is reflected in commit message for 6704c42a9ec79828057c4b2bcb813512d5e98f1e
